keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,336 @@
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This KEP generalizes the existing max-in-flight request handler in the
+apiserver to make more distinctions among requests and provide
+prioritization and fairness among the categories of requests.  An
+outline of the request handling in an apiserver can be found at
+https://speakerdeck.com/sttts/kubernetes-api-codebase-tour?slide=18 .
+
+## Motivation
+
+Today the apiserver has a simple mechanism for protectimg itself
+against CPU and memory overloads: max-in-flight limits for mutating
+and for readonly requests.  Apart from the distinction between
+mutating and readonly, no other distinctions are made among requests;
+consequently there can be undesirable scenarios where one subset of
+the request load crowds out other parts of the request load.  Also,
+the dispatching of these requests against two independent limits is
+not
+[work-conserving](https://en.wikipedia.org/wiki/Work-conserving_scheduler).
+
+### Goals
+
+Following are some bad scenarios that can happen today and which
+should be preventable when this KEP is in place.
+
+- Self-Maintenance crowded out.  Some requests are for system
+  self-maintenance, such as: node heartbeats, kubelet and kube-proxy
+  work on pods, services, secrets, etc involved in the system's
+  self-hosting, and leader elections for system controllers.  In an
+  overload scenario today there is no assurance of priority for these
+  self-maintenance requests.
+
+- Priority Inversions.  In the course of serving request A, there are
+  some other requests B spawned --- directly or indirectly.  One
+  example is requests B that arrive over a loopback connection ---
+  such as requests issued by an admission plugin (e.g., ResourceQuota)
+  or any client-ish code in a registry strategy.  Another example is
+  requests issued by an external server that itself is serving
+  call-outs from an apiserver (e.g., admission web-hooks).  Other
+  examples include requests from an aggregated apiserver to create
+  TokenReview and SubjectAccessReview objects.  Today it is possible
+  that the very load imposed by request A crowds out requests B
+  involved in serving A.
+
+- Garbage Collector crowded out.  The garbage collector should keep up
+  with the workload, but in an overload situation today this is not
+  assured to happen.
+
+- Deployment of Doom.  We had a situation where a bug in the
+  Deployment controller caused it to run amuck under certain
+  circumstances, issuing requests in a tight loop.  We would like
+  controller bugs to not take the whole system down.
+
+- Kubelet Amuck.  The controller that runs amuck might not be a
+  central singleton, it could be a kubelet, kube-proxy, or other
+  per-node or otherwise multiplied controller.  In such a situtation
+  we would like only the guilty individual to suffer, not all its
+  peers and the rest of the system.
+
+- Overbearing or buggy tenants.  In a multi-tenant scenario, we would
```

Comment:
I think this is a future goal, not a current goal.  I would be very happy with a system where this is probably doable at some point, but I think keeping the control-plane alive is required for MVP and idea of tenancy will bring with different slicing dimensions that are interesting, but not required to have significant benefit.

Reply:
I think this is isomorphic to the Kubelet Amuck.

Reply:
I agree this is only a future goal, however I think any design that solves our other goals will accidentally solve this, with only an additional slicing dimension necessary if anything.

Reply:
This is the goal that brought me here.  Given that we can meet it with no significant additional complexity, I think we should include it from the start.